itest: simple taproot channel activation test

hieblmi opened the pull request:
Ensures that https://github.com/lightningnetwork/lnd/issues/8065 is fixed.

This PR adds an integration test that ensures that a channel appears active if the initiating peer is disconnected from the channel peer during channel confirmation and reconnected after the channel confirmed.

Depends on,
- #8078

yyforyongyu commented on itest/lnd_open_channel_test.go:
```diff
@@ -768,3 +768,80 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
 	ht.CloseChannel(alice, chanPoint)
 }
+
+// testSimpleTaprootChannelActivation ensures that a simple taproot channel is
```
👍

Crypt-iQ commented on itest/lnd_open_channel_test.go:
```diff
@@ -768,3 +768,80 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+			Host:   bob.Cfg.P2PAddr(),
+		},
+	})
+
```
If you add a sleep here, the test passes. Because of that, I don't think this is a proper reproduction of the issue

hieblmi replied:
Ok, let me check again the manual steps I take in regtest to produce the error logs.

yyforyongyu commented on itest/lnd_open_channel_test.go:
```diff
@@ -768,3 +768,80 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// We disconnect Bob from Alice before the channel is confirmed
+	// on-chain. Our expectation is that once the channel is confirmed and
+	// Alice reconnects, the channel will show as active on Alice's side.
+	alice.RPC.DisconnectPeer(bob.PubKeyStr)
```
can use `ht.EnsureConnected(alice, bob)`

yyforyongyu commented on itest/lnd_open_channel_test.go:
```diff
@@ -768,3 +768,80 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	require.Len(ht, resp.Channels, 1)
+
+	// Now Alice reconnects to Bob.
+	alice.RPC.ConnectPeer(&lnrpc.ConnectPeerRequest{
```
same here, can use `ht.EnsureConnected(alice, bob` instead to assert they are connected.

yyforyongyu commented on itest/lnd_open_channel_test.go:
```diff
@@ -768,3 +768,80 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	resp = alice.RPC.ListChannels(&lnrpc.ListChannelsRequest{
+		ActiveOnly: true,
+	})
+	require.Len(ht, resp.Channels, 1)
```
can instead replace this check with `ht.AssertChannelExists(alice, chanPoint)`

yyforyongyu replied:
nit: also maybe we could add more methods here,
- `ht.AssertChannelActive`, which is basically the current `ht.AssertChannelExists`
- `ht.AssertChannelInactive`, which checks the channel exists and is inactive by [checking it here](https://github.com/lightningnetwork/lnd/blob/master/lntest/harness_assertion.go#L367)
- `ht.AssertChannelExists`, which only checks the channel exists

yyforyongyu commented on itest/lnd_open_channel_test.go:
```diff
@@ -768,3 +768,80 @@ func testFundingExpiryBlocksOnPending(ht *lntest.HarnessTest) {
+	// on-chain. Our expectation is that once the channel is confirmed and
+	// Alice reconnects, the channel will show as active on Alice's side.
+	alice.RPC.DisconnectPeer(bob.PubKeyStr)
+
```
so if we connect the nodes again here, we'd see this error in Alice's log,
```
2023-10-10 19:31:50.648 [WRN] HSWC: ChannelLink(66bea0e467d87ade4430bcf08d485dbdfaa40f092bb88337c057c34b1df4007c:0): error when syncing channel states: unable to init remote nonce: pending verification nonce is not set
2023-10-10 19:31:50.648 [ERR] HSWC: ChannelLink(66bea0e467d87ade4430bcf08d485dbdfaa40f092bb88337c057c34b1df4007c:0): failing link: unable to synchronize channel states: unable to init remote nonce: pending verification nonce is not set with error: unable to resume channel, recovery required
```

I think it indicates there's a connection attempt in your regtest environment - since this is a permanent connection, alice(zane) will try to reconnect to bob.

Crypt-iQ replied:
nice catch, I've now identified the root cause here

hieblmi replied:
Thanks for the hint @yyforyongyu. @Crypt-iQ, will complete this test in a bit.

hieblmi commented:
Thanks @yyforyongyu and @Crypt-iQ for the review and fix, I addressed suggestions and nits and the test now repros the error reliably. I also ran the test on top of https://github.com/lightningnetwork/lnd/pull/8078 which makes the test succeed.

yyforyongyu commented on lntest/harness_assertion.go:
```diff
@@ -377,6 +377,56 @@ func (h *HarnessTest) AssertChannelExists(hn *node.HarnessNode,
+		}
+
+		// Check whether the channel is active, exit early if it is.
+		if active && channel.Active {
```
think you can simplify these two `if`s as `channel.Active == active`.

yyforyongyu commented on lntest/harness_assertion.go:
```diff
@@ -377,6 +377,56 @@ func (h *HarnessTest) AssertChannelExists(hn *node.HarnessNode,
+// AssertChannelStatus asserts that a channel identified by the specified
+// channel point exists from the point-of-view of the node and that it is either
+// active or inactive depending on the value of the active parameter.
+func (h *HarnessTest) AssertChannelStatus(hn *node.HarnessNode,
```
let's make this private? Reason being it'd be easier for caller to finish the auto-complete in the editor.

hieblmi replied:
Do itests warrant an update of the release notes?

yyforyongyu replied:
Good question - atm no, but maybe we should think about it in future releases.

yyforyongyu reviewed (commented):
Looking good! Will approve this once the dependent PR is merged🏖

Crypt-iQ commented on lntest/harness_assertion.go:
```diff
@@ -377,6 +377,56 @@ func (h *HarnessTest) AssertChannelExists(hn *node.HarnessNode,
+		if channel.Active == active {
+			return nil
+		}
+		if !channel.Active == !active {
```
This `if` and the above `if` test the same thing so we can remove the second one

Crypt-iQ reviewed (commented):
One comment, then can land after dependent PR merged

Crypt-iQ commented:
Closing in favor of #8078
